### PR TITLE
Sizeable refactor and comms builder

### LINKED ISSUE #367: Comms Builder
Create an public interface to set up the comms layer that's intuitive and hides away some of the current complexities of setting up the various components required for the comms layer to run.

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

This should be the last of the large refactors on the comms layer. This PR introduces a CommsBuilder which is the  public interface for starting up the various comms services. 

What was done:
- Renamed Context to `ZmqContext`
- Added CommsBuilder 
- Control Service will now send a domain-level message type to the peer connection
- Refactored the generics to make them simpler to use - the only generic we need to know from the domain layer is 'what type of the MessageType'.
- Removed some message copies 
- Removed the global node identity static in favour of passing the node identity to the components which require it
- More sensible defaults to make the public api simpler
- Added test factories for connection manager
- Some docs

What needs to be completed:
- Tech debt (docs and tests)
- Comms builder will need some work but this should work for hooking things up

## Motivation and Context

Closes #367 

## How Has This Been Tested?

Existing tests pass (with  updates to accommodate breaking api changes)
Basic comms builder test

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [x] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [x] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [~] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/builder/builder.rs [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+            // OMP can requeue back onto itself
+            message_sink_address.clone(),
+            message_sink_address.clone(),
+            peer_manager,
+            connection_manager,
+        )
+        .map_err(CommsBuilderError::OutboundMessagePoolError)
+    }
+
+    pub fn make_inbound_message_service(
```
Comment (philipr-za) [inline comms/src/builder/builder.rs]: Nit: I prefer consistency, you were using acronyms before so rather use IMS here

File: comms/src/builder/builder.rs [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+    UncleanShutdown,
+}
+
+pub struct CommsServices<MType>
+where
+    MType: Serialize + DeserializeOwned,
+    MType: DispatchableKey,
+    MType: Clone,
+{
+    connection_manager: Arc<ConnectionManager>,
```
Comment (philipr-za) [inline comms/src/builder/builder.rs]: I think you are missing the OMS in this collection? It doesn't get started like the others but it is integral to using the comms layer.

Comment (sdbondi) [reply comms/src/builder/builder.rs]: Not needed in the current state of the code, but agree will be needed later - will leave it for another PR once we're there with the IMS dispatching changes

File: comms/src/builder/builder.rs (hunk 1 above)
Comment (sdbondi) [reply comms/src/builder/builder.rs]: Done

**Pull request merged**